tests/bluetooth/host/conn/src/main.c

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/fff.h>
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+
+#include <host_mocks/assert.h>
+
+#include "mocks/addr_internal.h"
+#include "mocks/att_internal.h"
+#include "mocks/bt_str.h"
+#include "mocks/buf_view.h"
+#include "mocks/hci_core.h"
+#include "mocks/id.h"
+#include "mocks/kernel.h"
+#include "mocks/l2cap_internal.h"
+#include "mocks/scan.h"
+#include "mocks/smp.h"
+#include "mocks/spinlock.h"
+#include "mocks/sys_clock.h"
+
+DEFINE_FFF_GLOBALS;
+
+static void fff_reset_rule_before(const struct ztest_unit_test *test, void *fixture)
+{
+	ADDR_INTERNAL_MOCKS_FFF_FAKES_LIST(RESET_FAKE);
+	ATT_INTERNAL_MOCKS_FFF_FAKES_LIST(RESET_FAKE);
+	BUF_VIEW_MOCKS_FFF_FAKES_LIST(RESET_FAKE);
+	HCI_CORE_MOCKS_FFF_FAKES_LIST(RESET_FAKE);
+	ID_MOCKS_FFF_FAKES_LIST(RESET_FAKE);
+	KERNEL_MOCKS_FFF_FAKES_LIST(RESET_FAKE);
+	L2CAP_INTERNAL_MOCKS_FFF_FAKES_LIST(RESET_FAKE);
+	SCAN_MOCKS_FFF_FAKES_LIST(RESET_FAKE);
+	SMP_MOCKS_FFF_FAKES_LIST(RESET_FAKE);
+	SPINLOCK_MOCKS_FFF_FAKES_LIST(RESET_FAKE);
+	SYS_CLOCK_MOCKS_FFF_FAKES_LIST(RESET_FAKE);
+}
+
+ZTEST_RULE(fff_reset_rule, fff_reset_rule_before, NULL);
+
+ZTEST_SUITE(conn, NULL, NULL, NULL, NULL, NULL);
+
+/*
+ * Test that bt_conn_le_create() returns -EINVAL if conn is not NULL and
+ * CONFIG_BT_CONN_CHECK_NULL_BEFORE_CREATE is enabled.
+ *
+ * The test must be compiled with and without CONFIG_BT_CONN_CHECK_NULL_BEFORE_CREATE
+ * to ensure that the -EINVAL error is returned only when this Kconfig option is enabled.
+ */
+ZTEST(conn, test_bt_conn_le_create_check_null_conn)
+{
+	bt_addr_le_t peer = {.a.val = {0x01}};
+	const struct bt_conn_le_create_param create_param = {
+		.options = BT_CONN_LE_OPT_NONE,
+		.interval = BT_GAP_SCAN_FAST_INTERVAL,
+		.window = BT_GAP_SCAN_FAST_WINDOW,
+		.interval_coded = 0,
+		.window_coded = 0,
+		.timeout = 100UL * MSEC_PER_SEC / 10,
+	};
+	struct bt_conn *conn;
+	int err;
+
+	/* Set conn to any non-NULL value to trigger the check. */
+	conn = (struct bt_conn *)1;
+
+	expect_assert();
+
+	err = bt_conn_le_create(&peer, &create_param, BT_LE_CONN_PARAM_DEFAULT, &conn);
+	/* err must not be -EINVAL if CONFIG_BT_CONN_CHECK_NULL_BEFORE_CREATE is not enabled,
+	 * otherwise -EAGAIN is returned.
+	 */
+#if defined(CONFIG_BT_CONN_CHECK_NULL_BEFORE_CREATE)
```

Comment:
`IS_ENABLED` checks if macro is 1 or not. If macro is defined to `y` through CMake `EXTRA_CPPFLAGS` or in `target_compile_options`, `IS_ENABLED` will be converted `0` but `if defined` will still compile the correct branch and the test will fail. So I'd like to keep it here in case someone (like me while implementing the test) occasionally sets the option to `y`.

Reply:
@PavelVPV isn't that just a recipe for really hard to diagnose bugs? Whatever the source for Kconfig options the resulting preprocessor macros should have the same values and semantics, so I'd focus on fixing that. Maybe there's a better way to do this? Isn't the normal way to put something like the following into `testcase.yaml`:
```yml
extra_configs:
  - CONFIG_FOO=y
```

Reply:
If I use `extra_configs`, I need to pull in a lot of dependencies, like `CONFIG_BT`, `CONFIG_BT_HCI`, etc.

Reply:
Removed `extra_args` and added `extra_configs`.